bamboo/unit_tests/test_unit_layer_sigmoid_binary_cross_entropy.py

```diff
@@ -0,0 +1,205 @@
+import tools
+
+# ==============================================
+# Objects for Python data reader
+# ==============================================
+# Note: The Python data reader imports this file as a module and calls
+# the functions below to ingest data.
+
+# Data
+# Note: Sigmoid cross entropy is not differentiable w.r.t. ground
+# truth at 0 and 1.
+np.random.seed(20191218)
+_samples = np.random.normal(size=(11,2,13)).astype(np.float32)
+_samples[:,1,:] = np.clip(_samples[:,1,:], 0.1, 0.9)
+
+# Sample access functions
+def get_sample(index):
+    return _samples[index].reshape(-1)
+def num_samples():
+    return _samples.shape[0]
+def sample_dims():
+    return (2*_samples.shape[-1],)
+
+# ==============================================
+# Setup LBANN experiment
+# ==============================================
+
+def setup_experiment(lbann):
+    """Construct LBANN experiment.
+
+    Args:
+        lbann (module): Module for LBANN Python frontend
+
+    """
+    trainer = lbann.Trainer()
+    model = construct_model(lbann)
+    data_reader = construct_data_reader(lbann)
+    optimizer = lbann.NoOptimizer()
+    return trainer, model, data_reader, optimizer
+
+def construct_model(lbann):
+    """Construct LBANN model.
+
+    Args:
+        lbann (module): Module for LBANN Python frontend
+
+    """
+
+    # Input data
+    # Note: Sum with weights layers so that gradient checking will
+    # verify that error signals are correct.
+    slice_size = _samples.shape[-1]
+    x0_weights = lbann.Weights(optimizer=lbann.SGD(),
+                               initializer=lbann.ConstantInitializer(value=0.0),
+                               name='input0_weights')
+    x1_weights = lbann.Weights(optimizer=lbann.SGD(),
+                               initializer=lbann.ConstantInitializer(value=0.0),
+                               name='input1_weights')
+    x_slice = lbann.Slice(lbann.Input(),
+                          slice_points=tools.str_list([0, slice_size, 2*slice_size]))
+    x0 = lbann.Sum(x_slice,
+                   lbann.WeightsLayer(weights=x0_weights, dims=str(slice_size)))
+    x1 = lbann.Sum(x_slice,
+                   lbann.WeightsLayer(weights=x1_weights, dims=str(slice_size)))
+    x0_lbann = x0
+    x1_lbann = x1
+
+    # Objects for LBANN model
+    obj = []
+    metrics = []
+    callbacks = []
+
+    # ------------------------------------------
+    # Data-parallel layout
+    # ------------------------------------------
+
+    # LBANN implementation
+    x0 = x0_lbann
+    x1 = x1_lbann
+    y = lbann.SigmoidBinaryCrossEntropy(x0, x1, data_layout='data_parallel')
+    z = lbann.L2Norm2(y)
+    obj.append(z)
+    metrics.append(lbann.Metric(z, name='data-parallel layout'))
+
+    # NumPy implementation
+    vals = []
+    for i in range(num_samples()):
+        x = get_sample(i).astype(np.float64)
+        x0 = x[:slice_size]
+        x1 = x[slice_size:]
+        y = -x1 * np.log1p(np.exp(-x0)) - (1-x1) * np.log1p(np.exp(x0))
```

Comment:
This doesnt exactly match x1 * log(1 + exp(-x0)) + (1 - x1) * (x0 + log(1 + exp(-x0)) from [TF](https://www.tensorflow.org/api_docs/python/tf/nn/sigmoid_cross_entropy_with_logits?version=stable)? Just checking, I could be wrong.

Reply:
It's mathematically the same. They rearranged stuff to avoid numerical errors, but this test isn't in a regime where that is a big concern.